### PR TITLE
Documentation: added suggested introduction to interrupt handlers

### PR DESCRIPTION
**Pull request by peterhinch**
File added docs/reference/isr_rules.rst
Regular questions in the forum indicated the need for a basic introduction to the topic of writing interrupt callbacks, and this proposed addition stems from this discussion
http://forum.micropython.org/viewtopic.php?f=6&t=1156
The text can be viewed here http://hinch.me.uk/html/reference/isr_rules.html

### CODE REVIEW

**Comment by dpgeorge** [comment]
Wow, thanks!  I'll have to read through it properly but it looks good.

"Why can't I allocate in an ISR" is a very common problem, and this should help (or at least provide a place to refer people to).

Eventually it'll be nice to have "soft ISR" where response is less fast but you can allocate on the heap (as per a discussion elsewhere); and/or actually be able to allocate during a hardware ISR; and/or OS-level threads.  But I still think this discussion about restrictions on ISRs is useful, especially for people new to embedded stuff, and especially because uPy uses a heap more than traditional C.

File: docs/reference/isr_rules.rst [hunk 1]
```diff
@@ -0,0 +1,270 @@
...
+    micropython.alloc_emergency_exception_buf(100)
+    class Foo(object):
+        def __init__(self, timer, led):
+            self.led = led
+            timer.callback(self.cb)
+        def cb(self, tim):
+            self.led.toggle()
+
+    red = Foo(pyb.Timer(4, freq=1), pyb.LED(1))
+    greeen = Foo(pyb.Timer(2, freq=0.8), pyb.LED(2))
```
Comment (dpgeorge) [inline docs/reference/isr_rules.rst]: I think there needs to be a paragraph here which describes the code above, what the interrupts are and what they call.  Would also be good to explain that self.cb is a bound method which is how different state is passed around.

File: docs/reference/isr_rules.rst [hunk 2]
```diff
@@ -0,0 +1,270 @@
...
+
+    def callback1(t):
+        global data, index
+        for x in range(5):
+            data[index] = pyb.rng() # simulate input
+            index += 1
+            if index >= ARRAYSIZE:
+                raise BoundsException('Array bounds exceeded')
+
+    tim4 = pyb.Timer(4, freq=100, callback = callback1)
```
Comment (dpgeorge) [inline docs/reference/isr_rules.rst]: No space around equals in keyword.

File: docs/reference/isr_rules.rst [hunk 3]
```diff
@@ -0,0 +1,270 @@
...
+
+    for loop in range(1000):
+        if index > 0:
+            irq_state = pyb.disable_irq()
+            for x in range(index):
```
Comment (dpgeorge) [inline docs/reference/isr_rules.rst]: Maybe a comment saying "this is the critical section".

File: docs/reference/isr_rules.rst [hunk 4]
```diff
@@ -0,0 +1,270 @@
...
+                print(data[x])
+            index = 0
+            pyb.enable_irq(irq_state)
+            print('loop {}'.format(loop))
+        pyb.delay(1)
+
+    tim4.callback(None)
+
+A critical section can comprise a single line of code and a single variable.
+`This <https://github.com/dhylands/upy-examples/blob/master/atomic.py>`_ example illustrates a subtle source
```
Comment (dpgeorge) [inline docs/reference/isr_rules.rst]: Let's try not to link to external entities.  I think the only thing relevant here is the `t.counter+=1` piece of code.

**Comment by dpgeorge** [comment]
After reading this interrupts sound pretty scary!  I really think we need to implement "soft interrupts" which would alleviate almost all problems outlined in this document.  But of course still allow hard interrupts when they are needed.

I think the text needs a section "Tips and recommended practices" which is a sort of TL/DR section stating things like:
- keep isrs short
- don't allocate heap memory (eg no list append, dict inserts)
- use preallocated bytearray (or array.array) objects to share data
- use disable/enable irq in the main code to protect critical sections

On a separate but related note, I will make a document describing things that do and don't allocate on the heap. That will help a lot to know what code you can write in a hard isr.

File: docs/reference/isr_rules.rst (hunk 4 above)
Comment (dhylands) [inline docs/reference/isr_rules.rst]: Feel free to copy the code into the documentation if that suits your purposes.

**Comment by peterhinch** [comment]
I've attempted to address these points. Is the concept of a "soft interrupt" detailed anywhere? It sounds intriguing.

**Comment by dpgeorge** [comment]
Thank you!  Merged in 521759ee18f94a0a55f1fcc8a1242da9cf92af55.

**Comment by peterhinch** [comment]
The new docs seem to lack links to this section. An entry under "The MicroPython language" and a link from tutorial section "5. The Switch, callbacks and interrupts" would help people find it.

**Comment by dpgeorge** [comment]
The link under "The MP Language" is already there, just maybe it wasn't built correctly when you looked at it.  And I just added a link from section 5 of the pyboard tutorial, in 1d191fdf036f753b04a9f3b9d5187cdc4aadf120.  Thanks for the tip!

**Pull request closed without merging**